state-sync/inter-component/event-notifications/src/lib.rs

```diff
@@ -0,0 +1,458 @@
+        events: Vec<ContractEvent>,
+    ) -> Result<(), Error>;
+
+    /// Force the subscription service to notify subscribers of a reconfiguration.
+    /// This is useful for forcing reconfiguration notifications even if no
+    /// reconfiguration event was processed (e.g., on startup).
+    async fn force_notify_reconfigurations(&mut self, version: Version) -> Result<(), Error>;
+}
+
+/// The subscription service offered by state sync, responsible for notifying
+/// subscribers of on-chain events.
+pub struct EventSubscriptionService {
+    // Event subscription registry
+    event_key_subscriptions: HashMap<EventKey, HashSet<SubscriptionId>>,
+    subscription_id_to_event_subscription: HashMap<SubscriptionId, EventSubscription>,
+
+    // Reconfig subscription registry
+    reconfig_subscriptions: HashMap<SubscriptionId, ReconfigSubscription>,
+
+    // Database to fetch on-chain configuration data
+    storage: Arc<RwLock<DbReaderWriter>>,
+
+    // Internal subscription ID generator
+    next_subscription_id: AtomicU64,
+}
+
+impl EventSubscriptionService {
+    pub fn new(storage: Arc<RwLock<DbReaderWriter>>) -> Self {
+        Self {
+            event_key_subscriptions: HashMap::new(),
+            subscription_id_to_event_subscription: HashMap::new(),
+            reconfig_subscriptions: HashMap::new(),
+            storage,
+            next_subscription_id: AtomicU64::new(0),
+        }
+    }
+
+    /// Returns an EventNotificationListener that can be monitored for
+    /// subscribed events. If an event key is subscribed to, it means the
+    /// EventNotificationListener will be sent a notification every time an
+    /// event with the matching key occurs on-chain. Note: if the notification
+    /// buffer fills up too quickly, older notifications will be dropped. As
+    /// such, it is the responsibility of the subscriber to ensure notifications
+    /// are processed in a timely manner.
+    pub fn subscribe_to_events(
+        &mut self,
+        event_keys: Vec<EventKey>,
+    ) -> Result<EventNotificationListener, Error> {
+        if event_keys.is_empty() {
+            return Err(Error::CannotSubscribeToZeroEventKeys);
+        }
+
+        let (notification_sender, notification_receiver) =
+            diem_channel::new(QueueStyle::KLAST, EVENT_NOTIFICATION_CHANNEL_SIZE, None);
+
+        // Create a new event subscription
+        let subscription_id = self.get_new_subscription_id();
+        let event_subscription = EventSubscription {
+            subscription_id,
+            notification_sender,
+            event_buffer: vec![],
+        };
+
+        // Store the new subscription
+        if let Some(old_subscription) = self
+            .subscription_id_to_event_subscription
+            .insert(subscription_id, event_subscription)
+        {
+            return Err(Error::UnknownEventSubscription(
+                subscription_id,
+                format!("{:?}", old_subscription),
+            ));
+        }
```

Comment:
That's kinda a weirdly named error.  Maybe, subscription already exists?  Because this would only happen if get_new_subscription_id gave the same ID twice right?

Also, maybe we should just crash at this point?  if we're throwing away a subscription that already exists, we're not in a good state

Reply:
yeah, this is like internal invariant of get_new_subscription_id?

Reply:
Yeah, this would be an invariant violation. I've updated the code to panic and removed the error type 😄